test/extensions/quic_listeners/quiche/integration/quic_http_integration_test.cc

```diff
@@ -218,5 +228,138 @@ TEST_P(QuicHttpIntegrationTest, TestDelayedConnectionTeardownTimeoutTrigger) {
+          fmt::format("listener.0.0.0.0_0.worker_{}.downstream_cx_active", i), 1u);
+      test_server_->waitForCounterEq(
+          fmt::format("listener.0.0.0.0_0.worker_{}.downstream_cx_total", i), 1u);
+    } else {
+      test_server_->waitForGaugeEq(fmt::format("listener.[__]_0.worker_{}.downstream_cx_active", i),
+                                   1u);
+      test_server_->waitForCounterEq(
+          fmt::format("listener.[__]_0.worker_{}.downstream_cx_total", i), 1u);
+    }
+  }
+  for (size_t i = 0; i < concurrency_; ++i) {
+    codec_clients[i]->close();
+  }
+#endif
+}
+
+TEST_P(QuicHttpIntegrationTest, MultipleQuicListenersNoBPF) {
+#ifdef __APPLE__
+  return;
+#endif
+  concurrency_ = 8;
+  set_reuse_port_ = true;
+  initialize();
+#ifdef SO_ATTACH_REUSEPORT_CBPF
+#define SO_ATTACH_REUSEPORT_CBPF_TMP SO_ATTACH_REUSEPORT_CBPF
+#undef SO_ATTACH_REUSEPORT_CBPF
+#endif
+  std::vector<IntegrationCodecClientPtr> codec_clients;
+  quic::QuicCryptoClientConfig::CachedState* cached = crypto_config_.LookupOrCreate(server_id_);
+  for (size_t i = 1; i <= concurrency_; ++i) {
+    // The BPF filter looks at the 1st byte of connection id in the packet
+    // header. And currently all QUIC versions support 8 bytes connection id. So
+    // create connections with the first 4 bytes of connection id different from each
+    // other so they should be evenly distributed.
+    cached->add_server_designated_connection_id(quic::test::TestConnectionId(i << 32));
+    codec_clients.push_back(makeHttpConnection(lookupPort("http")));
+  }
+  if (GetParam() == Network::Address::IpVersion::v4) {
+    test_server_->waitForCounterEq("listener.0.0.0.0_0.downstream_cx_total", 8u);
+  } else {
+    test_server_->waitForCounterEq("listener.[__]_0.downstream_cx_total", 8u);
+  }
+  // Even without BPF support, these connections should more or less distributed
+  // across different workers.
+  for (size_t i = 0; i < concurrency_; ++i) {
+    if (GetParam() == Network::Address::IpVersion::v4) {
+      EXPECT_LT(
+          test_server_->gauge(fmt::format("listener.0.0.0.0_0.worker_{}.downstream_cx_active", i))
+              ->value(),
+          8u);
+      EXPECT_LT(
+          test_server_->counter(fmt::format("listener.0.0.0.0_0.worker_{}.downstream_cx_total", i))
+              ->value(),
+          8u);
+    } else {
+      EXPECT_LT(
+          test_server_->gauge(fmt::format("listener.[__]_0.worker_{}.downstream_cx_active", i))
+              ->value(),
+          8u);
+      EXPECT_LT(
+          test_server_->counter(fmt::format("listener.[__]_0.worker_{}.downstream_cx_total", i))
+              ->value(),
+          8u);
+    }
+  }
+  for (size_t i = 0; i < concurrency_; ++i) {
+    codec_clients[i]->close();
+  }
+#ifdef SO_ATTACH_REUSEPORT_CBPF_TMP
+#define SO_ATTACH_REUSEPORT_CBPF SO_ATTACH_REUSEPORT_CBPF_TMP
+#endif
+}
+
+TEST_P(QuicHttpIntegrationTest, ConnectionMigration) {
+#if defined(SO_ATTACH_REUSEPORT_CBPF) && defined(__linux__)
```

Comment:
might as well put the TEST_P {} in this block too, so we don't create and run an empty test if it's not testable, yeah?